File: app/src/main/java/ezike/tobenna/myweather/data/local/LocalDataSourceImpl.java
package ezike.tobenna.myweather.data.local;

import org.threeten.bp.ZonedDateTime;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import ezike.tobenna.myweather.data.local.dao.WeatherDao;
import ezike.tobenna.myweather.data.local.entity.WeatherResponse;
import ezike.tobenna.myweather.provider.LocationProvider;

public class LocalDataSourceImpl implements LocalDataSource<WeatherResponse, LiveData<WeatherResponse>> {

    private final WeatherDao mWeatherDao;

    private LocationProvider mLocationProvider;

    @Inject
    LocalDataSourceImpl(LocationProvider locationProvider,
                        WeatherDao weatherDao) {
        mLocationProvider = locationProvider;
        mWeatherDao = weatherDao;
    }

    @Override
    public void save(WeatherResponse response) {
        mWeatherDao.insertWeatherResponse(response);
    }

    @Override
    public boolean hasLocationChanged(WeatherResponse response) {
        return mLocationProvider.isLocationChanged(response.getLocation());
    }

    @Override
    public boolean shouldFetch(WeatherResponse response) {
        ZonedDateTime locationTime = response.getLocation().getZonedDateTime();
        ZonedDateTime timeElapsed = ZonedDateTime.now().minusMinutes(30);
        return locationTime.isBefore(timeElapsed);
    }

    @Override
    public LiveData<WeatherResponse> get() {
        return mWeatherDao.getWeatherResponse();
    }
}

File: app/src/main/java/ezike/tobenna/myweather/data/local/LocalDataSourceImpl.kt
package ezike.tobenna.myweather.data.local

import ezike.tobenna.myweather.data.local.db.WeatherDao
import ezike.tobenna.myweather.data.model.WeatherResponse
import ezike.tobenna.myweather.provider.LocationProvider
import javax.inject.Inject

class LocalDataSourceImpl @Inject constructor(private val weatherDao: WeatherDao, private val locationProvider: LocationProvider)
    : LocalDataSource {

    override suspend fun save(weatherResponse: WeatherResponse?) {
        weatherDao.insert(weatherResponse)
    }

    override fun hasLocationChanged(weatherResponse: WeatherResponse): Boolean {
        return locationProvider.isLocationChanged(weatherResponse.weatherLocation);
    }

    override suspend fun getWeather(): WeatherResponse {
        return weatherDao.fetchWeather()
    }
}